### PR TITLE
Support target_project key for branch and link package step

### PR DESCRIPTION
**Pull request by krauselukas**
DO NOT MERGE: PR contains breaking changes, we need to communicate this before merging and deploying! But ready for reviews :)

### CODE REVIEW

**Comment by dmarcoux** [comment]
[Adapt workflow steps for push events on branches](https://trello.com/c/KUFzNTDw/1749-adapt-workflow-steps-for-push-events-on-branches)

File: src/api/app/models/workflow/step.rb [hunk 1]
```diff
@@ -15,7 +15,9 @@ def call(_options)
   end
 
   def target_project_name
-    "home:#{@token.user.login}:#{source_project_name}:PR-#{scm_webhook.payload[:pr_number]}"
+    return step_instructions[:target_project] if scm_webhook.push_event?
+
+    "#{step_instructions[:target_project]}:#{source_project_name}:PR-#{scm_webhook.payload[:pr_number]}"
```
Comment (hennevogel) [inline src/api/app/models/workflow/step.rb:20]: While we are at it we wanted to get off `source_project_name` here right? So more like

```ruby
"#{step_instructions[:target_project]}:#{scm_webhook.payload[:target_repository_full_name].split('/').join(':')}:PR-#{scm_webhook.payload[:pr_number]}"
```

Comment (krauselukas) [reply src/api/app/models/workflow/step.rb:20]: Alright, I overlooked that the naming scheme for pull request also changes, going to adapt it. Looking at the wiki now the naming looks a bit different `OBS:Server:Unstable:CI:PR-$MY_PR_NUMBER/obs-server`, but I like the way you described. So I would go for it

Comment (hennevogel) [reply src/api/app/models/workflow/step.rb:20]: I changed the wiki in the meanwhile ;-)

Comment (krauselukas) [reply src/api/app/models/workflow/step.rb]: @hennevogel Alright, so changed the code and test to use the new naming schema for PR's

**Comment by krauselukas** [comment]
Put back into WIP state because I have to check if permissions on the project are handled correctly

File: src/api/app/models/workflow/step.rb [hunk 2]
```diff
@@ -15,7 +15,15 @@ def call(_options)
   end
 
   def target_project_name
-    "home:#{@token.user.login}:#{source_project_name}:PR-#{scm_webhook.payload[:pr_number]}"
+    return step_instructions[:target_project] if scm_webhook.push_event?
+
+    pr_subproject_name = if scm_webhook.payload[:scm] == 'github'
```
Comment (vpereira) [inline src/api/app/models/workflow/step.rb:20]: isn't too optimistic to assume that this whole chain of commands will never fail?

Comment (krauselukas) [reply src/api/app/models/workflow/step.rb:20]: why? `target_repository_full_name` and `path_with_namespace` are always part of the payload, and there will always be one and only one `/`.

Comment (dmarcoux) [reply src/api/app/models/workflow/step.rb:20]: > there will always be one and only one `/`.

True :+1:. Slashes are replaced by dashes (`-`) in repository names.

File: src/api/app/models/workflow/step.rb [hunk 3]
```diff
@@ -15,7 +15,15 @@ def call(_options)
...
+                           scm_webhook.payload[:target_repository_full_name].split('/').first(2).join(':')
```
Comment (dmarcoux) [inline src/api/app/models/workflow/step.rb:21]: There is no need to split if you join the string again. Replacing the slashes for colon is much more straightforward.

```suggestion
                           scm_webhook.payload[:target_repository_full_name].gsub('/', ':')
```

File: src/api/app/models/workflow/step.rb [hunk 4]
```diff
@@ -15,7 +15,15 @@ def call(_options)
...
+                         else
+                           scm_webhook.payload[:path_with_namespace].split('/').first(2).join(':')
```
Comment (dmarcoux) [inline src/api/app/models/workflow/step.rb:23]: Same as above.

File: src/api/app/models/workflow/step.rb (hunk 3 above)
Comment (krauselukas) [reply src/api/app/models/workflow/step.rb]: done

File: src/api/app/models/workflow/step.rb (hunk 4 above)
Comment (krauselukas) [reply src/api/app/models/workflow/step.rb]: done

**Review by rubhanazeem — APPROVED** [review]
@krauselukas the changes look good to me. :+1:

**Comment by krauselukas** [comment]
Going to merge this now, and communicate the changes before the next deployment

**Pull request merged**